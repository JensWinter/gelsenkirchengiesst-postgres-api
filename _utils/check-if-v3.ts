export function urlContainsV3(url: string): boolean {
	return url.includes("v3");
}
