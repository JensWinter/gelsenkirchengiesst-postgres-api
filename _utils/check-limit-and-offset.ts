import { VercelRequest, VercelResponse } from "@vercel/node";
import { getEnvs } from "./envs";
const { SUPABASE_MAX_ROWS } = getEnvs();
export function checkLimitAndOffset(
	request: VercelRequest,
	response: VercelResponse
) {
	let { limit: limit_str, offset: offset_str } = <
		{ limit: string; offset: string }
	>request.query;

	if (!limit_str) {
		limit_str = `${SUPABASE_MAX_ROWS}`;
	}
	if (!offset_str) {
		offset_str = "0";
	}
	const limit = parseInt(limit_str, 10);
	const offset = parseInt(offset_str, 10);
	if (isNaN(limit)) {
		return response.status(400).json({ error: "limit needs to be a number" });
	}
	if (isNaN(offset)) {
		return response.status(400).json({ error: "offset needs to be a number" });
	}
	if (limit > SUPABASE_MAX_ROWS) {
		return response.status(400).json({
			error: `limit needs to be smaller than ${SUPABASE_MAX_ROWS}`,
		});
	}
}
