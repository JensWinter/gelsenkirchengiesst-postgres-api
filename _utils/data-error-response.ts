import { PostgrestError } from "@supabase/supabase-js";
import { VercelResponse } from "@vercel/node";

export function checkDataError({
	response,
	data,
	error,
	errorMessage,
}: {
	response: VercelResponse;
	data: unknown | null;
	error: PostgrestError | null;
	errorMessage: string;
}) {
	if (error) {
		return response.status(500).json({ error });
	}

	if (!data) {
		return response.status(500).json({ error: errorMessage });
	}
}
