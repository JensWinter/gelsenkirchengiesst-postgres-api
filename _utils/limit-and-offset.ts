import { VercelRequest, VercelResponse } from "@vercel/node";
import { getEnvs } from "./envs";
const { SUPABASE_MAX_ROWS } = getEnvs();
export function checkLimitAndOffset(
	request: VercelRequest,
	response: VercelResponse
) {
	let { limit: limit_str, offset: offset_str } = <
		{ limit: string; offset: string }
	>request.query;

	if (!limit_str) {
		limit_str = `${SUPABASE_MAX_ROWS}`;
	}
	if (!offset_str) {
		offset_str = "0";
	}
	const limit = parseInt(limit_str, 10);
	const offset = parseInt(offset_str, 10);
	if (isNaN(limit)) {
		return response.status(400).json({ error: "limit needs to be a number" });
	}
	if (isNaN(offset)) {
		return response.status(400).json({ error: "offset needs to be a number" });
	}
	if (limit > SUPABASE_MAX_ROWS) {
		return response.status(400).json({
			error: `limit needs to be smaller than ${SUPABASE_MAX_ROWS}`,
		});
	}
}

export function getLimitAndOffeset(query: {
	[key: string]: string | string[];
}): {
	limit: number;
	offset: number;
} {
	let { limit: limit_str, offset: offset_str } = <
		{ limit: string; offset: string }
	>query;

	if (!limit_str) {
		limit_str = `${SUPABASE_MAX_ROWS}`;
	}
	if (!offset_str) {
		offset_str = "0";
	}
	const limit = parseInt(limit_str, 10);
	const offset = parseInt(offset_str, 10);
	if (isNaN(limit) && isNaN(offset)) {
		return { limit: SUPABASE_MAX_ROWS, offset: 0 };
	}

	if (limit > SUPABASE_MAX_ROWS) {
		return { limit: SUPABASE_MAX_ROWS, offset };
	}

	if (offset < 0) {
		return { limit, offset: 0 };
	}
	return { limit, offset };
}
