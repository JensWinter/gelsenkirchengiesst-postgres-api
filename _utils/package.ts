import fs from "fs";
import path from "path";

export interface Package {
	[key: string]: string | number | Record<string, unknown> | undefined;
	version: string;
	name: string;
	homepage?: string;
	bugs?: {
		url: string;
		// [key: string]: string | number | Record<string, unknown>;
	};
}
export function getPackage(): Package {
	const pkgStr = fs.readFileSync(path.resolve("./package.json"), "utf8");
	const pkg = JSON.parse(pkgStr);
	return pkg;
}
