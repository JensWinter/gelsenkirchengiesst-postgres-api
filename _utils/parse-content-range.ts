import { getEnvs } from "./envs";

const { SUPABASE_SERVICE_ROLE_KEY, SUPABASE_ANON_KEY } = getEnvs();

/**
 * getRange that makes a fetch head request to get the content range. Accepts url and returns a promise with the range object.
 * @example
 * const { error, range } = await getRange("http://localhost:54321/trees",{
 * 	headers:{
 * 		Authorization: `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
 * 		apikey: SUPABASE_ANON_KEY,
 * 		"Range-Unit": "items",
 * 		Prefer: "count=exact"
 * 		}
 * 	});
 *
 */
export async function getRange(url: string) {
	let error: Error | null = null;
	let range: ContentRange | null = null;

	const response = await fetch(url, {
		headers: {
			Authorization: `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
			apikey: SUPABASE_ANON_KEY,
			"Range-Unit": "items",
			Prefer: "count=exact",
		},
		method: "HEAD",
	});

	if (!response.ok) {
		error = new Error("Response not ok");
	}
	if (response.status === 401) {
		error = new Error("Unauthorized");
	}

	const contentRange = response.headers.get("content-range");
	if (contentRange === null) {
		error = new Error("No content-range header");
	} else {
		const parsedRange = parseContentRange(contentRange);
		if (parsedRange === null) {
			error = new Error("Invalid content-range header");
		} else {
			range = parsedRange;
		}
	}

	return { error, range };
}

/**
 * ContentRange if a value could not be parsed to a number it is set to -1
 */
export interface ContentRange {
	start: number;
	end: number;
	total: number;
}

export function parseContentRange(str: string): ContentRange | null {
	if (!str) {
		return null;
	}
	const arr = str.split("/");
	if (arr.length !== 2) {
		return null;
	}
	const rangeRegex = /^(?<start>\d+|\*)-(?<end>\d+|\*)$/;
	const totalRegex = /^(?<total>\d+|\*|0)$/;
	const rangeMatch = arr[0].match(rangeRegex);
	const totalMatch = arr[1].match(totalRegex);
	const start =
		rangeMatch?.groups?.start === "*"
			? -1
			: parseInt(rangeMatch?.groups?.start ?? "-1");
	const end =
		rangeMatch?.groups?.end === "*"
			? -1
			: parseInt(rangeMatch?.groups?.end ?? "-1");
	const total =
		totalMatch?.groups?.total === "*"
			? -1
			: parseInt(totalMatch?.groups?.total ?? "-1");
	if (isNaN(start) || isNaN(end) || isNaN(total)) {
		return null;
	}

	return { start, end, total };
}
