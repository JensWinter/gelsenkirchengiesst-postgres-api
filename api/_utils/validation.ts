import Ajv from "ajv";

const ajv = new Ajv(); // options can be passed, e.g. {allErrors: true}

export interface AjvSchema {
	type: string;
	properties: {
		[key: string]: unknown;
	};
	required: string[];
	additionalProperties: boolean;
}

export const testSchema: AjvSchema = {
	type: "object",
	properties: {
		foo: { type: "integer" },
		bar: { type: "string" },
	},
	required: ["foo"],
	additionalProperties: false,
};

export const waterSchema: AjvSchema = {
	type: "object",
	properties: {
		uuid: { type: "string" },
		tree_id: { type: "string" },
		username: { type: "string" },
		timestamp: { type: "string" },
		amount: { type: "number" },
		queryType: { type: "string" },
	},
	required: ["uuid", "tree_id", "username", "timestamp", "amount"],
	additionalProperties: false,
};

export const adoptSchema: AjvSchema = {
	type: "object",
	properties: {
		uuid: { type: "string" },
		tree_id: { type: "string" },
		queryType: { type: "string" },
	},
	required: ["uuid", "tree_id"],
	additionalProperties: false,
};

export const unadoptSchema: AjvSchema = {
	type: "object",
	properties: {
		uuid: { type: "string" },
		tree_id: { type: "string" },
		queryType: { type: "string" },
	},
	required: ["uuid", "tree_id"],
	additionalProperties: false,
};

export const unwaterSchema: AjvSchema = {
	type: "object",
	properties: {
		watering_id: { type: "number" },
		uuid: { type: "string" },
		tree_id: { type: "string" },
		queryType: { type: "string" },
	},
	required: ["uuid", "tree_id", "watering_id"],
	additionalProperties: false,
};

export { ajv };

export const validate = (body: Record<string, unknown>, schema: AjvSchema) => {
	const validate = ajv.compile(schema);
	const valid = validate(body);

	if (!valid) {
		return [false, validate.errors];
	}
	return [true, null];
};
