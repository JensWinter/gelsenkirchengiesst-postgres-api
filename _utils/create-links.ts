import { ContentRange } from "./parse-content-range";

// a function that create links for pagination based limit and offset and range
export function createLinks({
	limit,
	offset,
	range,
	type,
	method,
	requestUrl,
}: {
	limit: number;
	offset: number;
	range: ContentRange | null;
	type: string;
	method: string;
	requestUrl: string;
}) {
	const params = new URLSearchParams(requestUrl.split("?")[1]);
	// remove limit and offset from searchParams since we pass them as arguments
	// and they get some precomputing for SUPABASE_MAX_ROWS
	params.delete("limit");
	params.delete("offset");
	const paramsString =
		params.toString().length > 0 ? "&" + params.toString() : "";
	const links: Record<string, string> = {};
	if (range) {
		// If offset is greater than 0, generate a previous link
		const isPrevLink = offset > 0;
		if (isPrevLink) {
			links.prev = `/${method.toLowerCase()}/${type}?limit=${limit}&offset=${
				offset - limit < 0 ? 0 : offset - limit
			}${params ? paramsString : ""}`;
		}
		// If the offset is less than the total number of items minus 1, generate a next link
		const isNextLink = offset + limit < range.total - 1;
		if (isNextLink) {
			links.next = `/${method.toLowerCase()}/${type}?limit=${limit}&offset=${
				offset + limit
			}${params ? paramsString : ""}`;
		}
	}
	return links;
}
