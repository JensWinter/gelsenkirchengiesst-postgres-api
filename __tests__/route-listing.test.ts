import { listRoutes } from "../_utils/routes-listing";
import { paramsToObject, validate } from "../_utils/validation";
const getRoutesList = listRoutes("GET");
const postRoutesList = listRoutes("POST");
const deleteRoutesList = listRoutes("DELETE");
describe("route listing", () => {
	test("should validate urlSearchParams", () => {
		const params = paramsToObject("uuid=1234&limit=10&offset=0");
		const [valid, _validationErrors] = validate(
			params,
			getRoutesList.routes.wateredandadopted.schema
		);

		expect(valid).toBe(false);
	});

	test("should list all DELETE routes", () => {
		expect(deleteRoutesList).toMatchSnapshot();
	});
	test("should list all POST routes", () => {
		expect(postRoutesList).toMatchSnapshot();
	});

	test("should list all the GET routes", async () => {
		expect(getRoutesList).toMatchSnapshot();
	});
});
