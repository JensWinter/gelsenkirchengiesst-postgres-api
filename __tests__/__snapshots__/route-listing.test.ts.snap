// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route listing should list all DELETE routes 1`] = `
{
  "method": "DELETE",
  "routes": {
    "unadopt": {
      "schema": {
        "additionalProperties": false,
        "properties": {
          "queryType": {
            "type": "string",
          },
          "tree_id": {
            "type": "string",
          },
          "uuid": {
            "type": "string",
          },
        },
        "required": [
          "uuid",
          "tree_id",
        ],
        "type": "object",
      },
      "url": "delete/unadopt",
    },
    "unwater": {
      "schema": {
        "additionalProperties": false,
        "properties": {
          "queryType": {
            "type": "string",
          },
          "tree_id": {
            "type": "string",
          },
          "uuid": {
            "type": "string",
          },
          "watering_id": {
            "type": "number",
          },
        },
        "required": [
          "uuid",
          "tree_id",
          "watering_id",
        ],
        "type": "object",
      },
      "url": "delete/unwater",
    },
  },
}
`;

exports[`route listing should list all POST routes 1`] = `
{
  "method": "POST",
  "routes": {
    "adopt": {
      "schema": {
        "additionalProperties": false,
        "properties": {
          "queryType": {
            "type": "string",
          },
          "tree_id": {
            "type": "string",
          },
          "uuid": {
            "type": "string",
          },
        },
        "required": [
          "uuid",
          "tree_id",
        ],
        "type": "object",
      },
      "url": "post/adopt",
    },
    "water": {
      "schema": {
        "additionalProperties": false,
        "properties": {
          "amount": {
            "type": "number",
          },
          "queryType": {
            "type": "string",
          },
          "timestamp": {
            "type": "string",
          },
          "tree_id": {
            "type": "string",
          },
          "username": {
            "type": "string",
          },
          "uuid": {
            "type": "string",
          },
        },
        "required": [
          "uuid",
          "tree_id",
          "username",
          "timestamp",
          "amount",
        ],
        "type": "object",
      },
      "url": "post/water",
    },
  },
}
`;
